### PR TITLE
Observers and Filters implementation

### PR DESCRIPTION
**Pull request by santanusinha**

### CODE REVIEW

File: src/main/java/io/appform/dropwizard/sharding/observers/internal/TerminalTransactionObserver.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+import io.appform.dropwizard.sharding.execution.TransactionExecutionContext;
+import io.appform.dropwizard.sharding.observers.TransactionObserver;
+
+import java.util.function.Supplier;
+
+/**
+ *
+ */
+public class TerminalTransactionObserver extends TransactionObserver {
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/sharding/observers/internal/TerminalTransactionObserver.java:11]: What does this observer signify ?

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java [hunk 2]
```diff
@@ -142,23 +147,48 @@ protected String name() {
...
         sessionFactories = shardBundles.stream().map(HibernateBundle::getSessionFactory).collect(Collectors.toList());
         this.shardingOptions = getShardingOptions(configuration);
         environment.admin().addTask(new BlacklistShardTask(shardManager));
         environment.admin().addTask(new UnblacklistShardTask(shardManager));
         healthCheckManager.manageHealthChecks(getConfig(configuration).getBlacklist(), environment);
+
+        setupObservers();
+    }
+
+    public void registerObserver(final TransactionObserver observer) {
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java:169]: Should this method be final ?

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java [hunk 3]
```diff
@@ -142,23 +147,48 @@ protected String name() {
...
+        if (null == observer) {
+            return;
+        }
+        this.observers.add(observer);
+        log.info("Registered observer: " + observer.getClass().getSimpleName());
     }
 
+    public void registerListener(final TransactionListener listener) {
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java:177]: Should this method be final ?

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java [hunk 4]
```diff
@@ -142,23 +147,48 @@ protected String name() {
...
+        if (null == listener) {
+            return;
+        }
+        this.listeners.add(listener);
+        log.info("Registered listener: " + listener.getClass().getSimpleName());
+    }
+
+    public void registerFilter(final TransactionFilter filter) {
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java:185]: Should this method be final ?

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java [hunk 5]
```diff
@@ -51,18 +52,13 @@
...
 
 import javax.persistence.Entity;
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.List;
-import java.util.Map;
-import java.util.Objects;
-import java.util.Set;
+import java.util.*;
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java:61]: Usually using wildcard imports is not considered a good practice. SonarQube also raises code smell for this

File: src/main/java/io/appform/dropwizard/sharding/observers/TransactionObserver.java [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    protected TransactionObserver(TransactionObserver next) {
+        this.next = next;
+    }
+
+    public abstract <T> T execute(final TransactionExecutionContext context, Supplier<T> supplier);
+
+    public TransactionObserver setNext(final TransactionObserver next) {
+        this.next = next;
+        return this;
+    }
```
Comment (r0goyal) [inline src/main/java/io/appform/dropwizard/sharding/observers/TransactionObserver.java:27]: This can be made final

**Review by r0goyal — CHANGES_REQUESTED** [review]

File: src/main/java/io/appform/dropwizard/sharding/observers/internal/TerminalTransactionObserver.java (hunk 1 above)
Comment (santanusinha) [reply src/main/java/io/appform/dropwizard/sharding/observers/internal/TerminalTransactionObserver.java:11]: This basically marks the end point of the observer chain. Not strictly necessary, but decided to have this in place so as to have a clean marker.

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java (hunk 2 above)
Comment (santanusinha) [reply src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java]: Done

File: src/main/java/io/appform/dropwizard/sharding/observers/TransactionObserver.java (hunk 6 above)
Comment (santanusinha) [reply src/main/java/io/appform/dropwizard/sharding/observers/TransactionObserver.java]: Done

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java (hunk 5 above)
Comment (santanusinha) [reply src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java]: Done

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java (hunk 4 above)
Comment (santanusinha) [reply src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java]: Done

File: src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java (hunk 3 above)
Comment (santanusinha) [reply src/main/java/io/appform/dropwizard/sharding/DBShardingBundleBase.java]: Done

File: src/test/java/io/appform/dropwizard/sharding/LegacyDBShardingBundleWithListenerTest.java [hunk 7]
```diff
@@ -1,14 +1,8 @@
...
 import io.appform.dropwizard.sharding.config.ShardedHibernateFactory;
-import io.appform.dropwizard.sharding.dao.listeners.OrderItemTestListenerFactory;
-import io.appform.dropwizard.sharding.dao.listeners.TestListenerFactory;
 import io.appform.dropwizard.sharding.dao.testdata.entities.Order;
 import io.appform.dropwizard.sharding.dao.testdata.entities.OrderItem;
-import lombok.val;
-import org.junit.Assert;
-import org.junit.Test;
 
 public class LegacyDBShardingBundleWithListenerTest extends DBShardingBundleTestBase {
```
Comment (Kanikakhetawat) [inline src/test/java/io/appform/dropwizard/sharding/LegacyDBShardingBundleWithListenerTest.java:7]: BalancedDBShardingBundleWithListenerTest and this test class can be removed

Comment (santanusinha) [reply src/test/java/io/appform/dropwizard/sharding/LegacyDBShardingBundleWithListenerTest.java]: done

**Review by r0goyal — APPROVED** [review]

**Pull request merged**